src/scenic/core/lazy_eval.py

```diff
@@ -13,174 +13,195 @@
-	'__pow__', '__rpow__',
-	'__round__',
-	'__len__',
-	'__getitem__'
+    '__neg__',
+    '__pos__',
+    '__abs__',
+    '__lt__', '__le__',
+    '__eq__', '__ne__',
+    '__gt__', '__ge__',
+    '__add__', '__radd__',
+    '__sub__', '__rsub__',
+    '__mul__', '__rmul__',
+    '__truediv__', '__rtruediv__',
+    '__floordiv__', '__rfloordiv__',
+    '__mod__', '__rmod__',
+    '__divmod__', '__rdivmod__',
+    '__pow__', '__rpow__',
+    '__round__',
+    '__len__',
+    '__getitem__'
 ]
 def makeDelayedOperatorHandler(op):
-	def handler(self, *args):
-		props = self._requiredProperties.union(*(requiredProperties(arg) for arg in args))
-		def value(context):
-			subvalues = (valueInContext(arg, context) for arg in args)
-			return getattr(self.evaluateIn(context), op)(*subvalues)
-		return DelayedArgument(props, value, _internal=True)
-	return handler
+    def handler(self, *args):
+        props = set(self._requiredProperties).union(*(requiredProperties(arg) for arg in args))
+        def value(context, modifying):
+            subvalues = (valueInContext(arg, context, modifying) for arg in args)
+            return getattr(self.evaluateIn(context), op)(*subvalues)
+        return DelayedArgument(props, value, _internal=True)
+    return handler
 for op in allowedOperators:
-	setattr(DelayedArgument, op, makeDelayedOperatorHandler(op))
+    setattr(DelayedArgument, op, makeDelayedOperatorHandler(op))
 
 def makeDelayedFunctionCall(func, args, kwargs):
-	"""Utility function for creating a lazily-evaluated function call."""
-	props = set().union(*(requiredProperties(arg)
-	                      for arg in itertools.chain(args, kwargs.values())))
-	def value(context):
-		subvalues = (valueInContext(arg, context) for arg in args)
-		kwsubvals = { name: valueInContext(arg, context) for name, arg in kwargs.items() }
-		return func(*subvalues, **kwsubvals)
-	return DelayedArgument(props, value, _internal=True)
-
-def valueInContext(value, context):
-	"""Evaluate something in the context of an object being constructed."""
-	if needsLazyEvaluation(value):
-		return value.evaluateIn(context)
-	else:
-		return value
+    """Utility function for creating a lazily-evaluated function call."""
+    props = set().union(*(requiredProperties(arg)
+                          for arg in itertools.chain(args, kwargs.values())))
+    def value(context, modifying):
+        subvalues = (valueInContext(arg, context, modifying) for arg in args)
+        kwsubvals = { name: valueInContext(arg, context, modifying) for name, arg in kwargs.items() }
+        return func(*subvalues, **kwsubvals)
+    return DelayedArgument(props, value, _internal=True)
+
+def valueInContext(value, context, modifying={}):
+    """Evaluate something in the context of an object being constructed."""
+    if not needsLazyEvaluation(value):
+        return value
+    elif isinstance(value, LazilyEvaluable):
+        return value.evaluateIn(context, modifying)
+    elif isinstance(value, dict):
```

Comment:
Let's discuss this -- it seems fragile to put special cases for such types in all of these functions (`valueInContext`, `requiredProperties`, and `needsLazyEvaluation`). Maybe it would be better to have a `toLazyValue` function similar to `toDistribution` which produced a `DelayedArgument` for `dict`s, etc. containing lazy values.